lib/3scale/backend/metric.rb

```diff
@@ -162,6 +162,17 @@ def descendants(service_id, metric_name)
           end
         end
 
+        # Returns the "ascendants" of a metric, that is, its parents,
+        # grandparents, etc. The "ascendants" of a metric are its parents plus
+        # the ascendants of each of them.
+        def ascendants(service_id, metric_name)
+          parents_of_metric = parents(service_id, [metric_name])
+
+          parents_of_metric.reduce(parents_of_metric) do |acc, parent|
+            acc + ascendants(service_id, parent)
+          end
```

Comment:
Have you considered using `parents_of_metric.each` and pushing to it as you go for a more readable and performant version avoiding recursion?

Reply:
Not sure I follow. Can you expand a bit?

Reply:
:trollface:

Reply:
This is the exact same code as `descendants` - have you checked out how to DRY this up?

Reply:
I haven't come up with anything that makes the code more readable.

Reply:
This makes me feel uneasy, but oh well.